src/Featurizers/Components/GrainEstimatorImpl.h

```diff
@@ -0,0 +1,642 @@
+    // (I believe that that is a reasonable assumption).
+    //
+    // To check if the `Estimator` has created an `Annotation`, we will capture the size of the `AnnotationMaps`
+    // before `complete_training` and after `complete_training` is invoked. Changes in size indicate that an
+    // `Annotation` has been created (note that this isn't safe if multiple threads are simultaneously
+    // modifying the `AnnotationMap` (but that problem will go away when we switch to contexts)).
+    //
+    // If a new `Annotation` is detected, we will remove it from the global `AnnotationMaps` and add it to our
+    // internal `AnnotationMap`. Once complete, we will check for the presence of added `Annotations` and
+    // create a grain-based `Annotation` that contains all the grain-specific `Annotations`.
+    //
+    // TODO: All of this code needs to change when we switch from using `AnnotationMaps` in favor of graph
+    //       contexts.
+
+    Sizes const                             sizes(
+        [this](void) {
+            Sizes                           result;
+
+            result.reserve(_pAllColumnAnnotations->size());
+
+            for(AnnotationMap const &map: *_pAllColumnAnnotations)
+                result.emplace_back(map.size());
+
+            return result;
+        }()
+    );
+
+    ThisAnnotationMap                       newAnnotations;
+    size_t                                  colIndex(0);
+
+    for(auto & kvp : _estimators) {
+        kvp.second.complete_training();
+
+        bool                                addedNewAnnotation(false);
+
+        for(size_t i = 0; i < _pAllColumnAnnotations->size(); ++i) {
+            assert(_pAllColumnAnnotations->size() == sizes.size());
+            assert(i < _pAllColumnAnnotations->size());
+
+            AnnotationMap &                 map((*_pAllColumnAnnotations)[i]);
+            size_t const &                  size(sizes[i]);
+
+            if(map.size() != size) {
+                assert(map.size() == size + 1);
+
+                if(addedNewAnnotation)
+                    throw std::runtime_error("Unexpected AnnotationMap insertion (duplicate)");
+
+                if(newAnnotations.empty() == false && colIndex != i)
+                    throw std::runtime_error("Unexpected AnnotationMap insertion (different column)");
+
+                colIndex = i;
+
+                // If here, something was added. We expect that Annotation to be associated with the grain-based Estimator.
+                AnnotationMap::iterator     iter(map.find(kvp.second.Name));
+
+                if(iter == map.end())
+                    throw std::runtime_error("Unexpected AnnotationMap insertion (different Estimator)");
+
+                // There should be only one entry
+                if(iter->second.size() != 1)
+                    throw std::runtime_error("Unexpected AnnotationMap size");

```

Comment:
Can an estimator not add more then one annotation? Or if that happens is it probably because it wasn't decomposed enough?

Reply:
A templated featurizer class may lead to multiple concrete templates for a particular featurizer (for example, a featurizer configured to train on the first 10 items vs. a featurizer that is configured to train on all of the items). This check makes sure that an annotation from only 1 concrete template of this particular featurizer was added.